lambda/matching.ml

```diff
@@ -2959,27 +2956,20 @@ let combine_constructor loc arg pat_env cstr partial ctx def
                       | Some fail_act, Some a ->
                           Some (test_isint a fail_act)
                     in
-                    let numconsts, numblocks =
-                      let bounds = match vd.vd_max_values with
-                        | Some bounds -> bounds
-                        | None ->
-                          (* The values are not filled e.g in case a type is
-                             imported from a foreign cmi interface; we need to
-                             recompute them here *)
-                          if vd.vd_unboxed = 0 then
-                            Max vd.vd_consts, Max vd.vd_nonconsts
-                          else
-                            Typedecl_unboxed.Head_shape.max_val_of_cstr_descr
-                              pat_env cstr
-                      in match bounds with
-                        | Max imm, Max tag -> imm, tag
-                        | Unbounded, _ | _, Unbounded ->
-                            invalid_arg "Matching.combine_constructors"
+                    let sw_numconsts =
+                      match Typedecl.cstr_max_imm_value pat_env cstr with
+                      | Some n -> n + 1
+                      | None -> 1 (* XXX *)
```

Comment:
We discussed this. The problem is that  `simplif.ml` assumes that if there is a fail-action, then one of the two cases is partial. With this code the assumption is wrong when `fail_opt = None` but `any_const` is not (so the "merged" fail_opt is Some), and we cover all nonconst cases.

So in theory we should put `max_int` here if there is an `any_const`, but this would blow up compilation later that creates an array of this size.

`inline_lazy_force_switch` has the same issue, there solution is to do a `Pisint` test first and then a `switch` (with `sw_numconsts = 0`, `consts = []`) only in the block case. We can do the same here:
- if `any_const = None`, compile as in the current PR
- if `fail_opt = None` and `any_const = Some`, instead of defining a new `fail_opt` we should first generate a `Pisint` test, send to the `any_const` action in the then branch, and generate a switch only in the else branch
- if `fail_opt = Some` and `any_const = Some`, we can either generate a Pisint first (as above) or generate code as in the current PR (probably generate nicer code).

The solution of using `sw_numconsts = 1` may be unsafe (we are not sure) if the compiler then assumes that the immediate input may only be `0`, not any immediate.